The `ropensnp` package interacts with the API services of [OpenSNP](http://opensnp.org/). No API key is needed. 

This tutorial will go through three use cases to demonstrate the kinds of things possible in `ropensnp`.

+ Get genotype data for all users at a particular snp.
+ Get genotype data for one or multiple users.
+ Get phenotype data for one or multiple users.
+ Get openSNP user information.

***
### Load package from GitHub
```{r loadpacakge}
# install_github("ropensnp", "ropensci")
library(ropensnp)
```

***

#### Get genotype data for all users at a particular snp.
```{r allgensnp}
head( allgensnp('rs7412', df=TRUE)[,1:4] ) # minimal output to pretty presentation
```

***

#### Get genotype data for one or multiple users.
```{r genotypes}
genotypes(snp='rs9939609', userid=1)
head( genotypes('rs9939609', userid='1,6,8', df=TRUE)[,1:4] )
```

***

### Get phenotype data for one or multiple users.
```{r phenotypes}
dat <- phenotypes(userid='1,6,8', df=TRUE)
lapply(dat, head)
head( phenotypes(userid='1-8', df=TRUE)[[1]] )

library(plyr)
head( ldply(phenotypes(userid='1-8', df=TRUE)) )
```


***

### Get openSNP user information.
```{r users}
# just the list
data <- users(df=FALSE)
data[1:2]

# get a data.frame of the users data
data <- users(df=TRUE)
head( data[[1]][,c(1,5)] ) # users with links to genome data
head( data[[2]] ) # users without links to genome data
```


***

### Download genotype data for a user from 23andme or other repo.
```{r fetch_genotypes}
data <- users(df=TRUE) # get a data.frame of the users data
mydata <- fetch_genotypes(url = data[[1]][1,"genotypes.download_url"], rows=15)
mydata
```


***

### Get all known variations and all users sharing that phenotype for one phenotype(-ID).
```{r phenotypes_byid}
phenotypes_byid(phenotypeid=12, return_ = 'desc')
phenotypes_byid(phenotypeid=12, return_ = 'knownvars')
phenotypes_byid(phenotypeid=12, return_ = 'users')[1:3,]
```


***

### Get all SNPedia/PLos/Mendeley-annotations for a specific SNP.
```{r annotations}
annotations(snp = 'rs7903146', output = 'metadata') # get just the metadata
annotations(snp = 'rs7903146', output = 'plos')[,c(1,4)] # just from plos
annotations(snp = 'rs7903146', output = 'snpedia') # just from snpedia
dat <- annotations(snp = 'rs7903146', output = 'all') # get all annotations
str(dat)
```


***

### Get all phenotypes, their variations, and how many users have data available for a given phenotype.
```{r allphenotypes}
# Get all data
head( allphenotypes(df = TRUE) )

# Output a list, then call the characterisitc of interest by 'id' or 'characteristic'
datalist <- allphenotypes()
names(datalist)[1:4] # get list of all characteristics you can call, just a sample here
head( datalist[["ADHD"]] ) # get data.frame for 'ADHD'
datalist[c("mouth size","SAT Writing")] # get data.frame for 'ADHD'
```